spec/hotel_spec.rb

```diff
@@ -0,0 +1,371 @@
+      expect(@hotel.see_reservations_by_date(Date.new(2019, 3, 12)).length).must_equal 3
+      expect(@hotel.see_reservations_by_date(Date.new(2019, 3, 13)).length).must_equal 0
+    end
+  end
+
+  describe "see_available_rooms_by_date method" do
+    before do
+      @hotel = HotelSystem::Hotel.new
+      @room = @hotel.all_rooms[0]
+
+      @hotel.reserve_room(
+        @room,
+        Date.new(2019, 3, 11),
+        Date.new(2019, 3, 14),
+        "Sam"
+      )
+      @hotel.reserve_room(
+        @room,
+        Date.new(2019, 3, 11),
+        Date.new(2019, 3, 14),
+        "Sam"
+      )
+      @hotel.reserve_room(
+        @room,
+        Date.new(2019, 3, 11),
+        Date.new(2019, 3, 14),
+        "Sam"
+      )
+    end
+
+    it "returns an array" do
+      expect(@hotel.see_available_rooms_by_date(Date.new(2019, 3, 11), Date.new(2019, 3, 14))).must_be_kind_of Array
+    end
+
+    it "returns the correct rooms" do
+      expect(@hotel.see_available_rooms_by_date(Date.new(2019, 3, 11), Date.new(2019, 3, 14)).length).must_equal 17
+      expect(@hotel.see_available_rooms_by_date(Date.new(2019, 3, 11), Date.new(2019, 3, 14)).first.number).must_equal 4
+      expect(@hotel.see_available_rooms_by_date(Date.new(2019, 3, 11), Date.new(2019, 3, 14)).last.number).must_equal 20
+      expect(@hotel.see_available_rooms_by_date(Date.new(2019, 3, 11), Date.new(2019, 3, 14))[0].number).must_equal 4
+    end
+  end
+
+  describe "reserve_room method" do
+    let (:hotel) { HotelSystem::Hotel.new }
+    let (:room) { hotel.all_rooms[0] }
+
+    it "can add the reservation to the @all_reservations array" do
+      hotel.reserve_room(
+        room,
+        Date.new(2019, 3, 11),
+        Date.new(2019, 3, 14),
+        "Sam"
+      )
+      hotel.reserve_room(
+        room,
+        Date.new(2019, 3, 11),
+        Date.new(2019, 3, 14),
+        "Sam"
+      )
+      hotel.reserve_room(
+        room,
+        Date.new(2019, 3, 11),
+        Date.new(2019, 3, 14),
+        "Sam"
+      )
+
+      expect(hotel.all_reservations.length).must_equal 3
+      expect(hotel.all_reservations[0]).must_be_kind_of HotelSystem::Reservation
+      expect(hotel.all_reservations[1]).must_be_kind_of HotelSystem::Reservation
+      expect(hotel.all_reservations[2]).must_be_kind_of HotelSystem::Reservation
+    end
+
+    it "can add the reservation to the assigned room's reservation array" do
+      hotel.reserve_room(
+        room,
+        Date.new(2019, 3, 11),
+        Date.new(2019, 3, 14),
+        "Sam"
+      )
+      expect(room.reservations.length).must_equal 1
+      expect(room.reservations[0]).must_be_kind_of HotelSystem::Reservation
+    end
+
+    it "can assign a room, that doesn't have other conflicting reservations, to the reservation if the selected room is not available" do
+      hotel.reserve_room(
+        room,
+        Date.new(2019, 3, 10),
+        Date.new(2019, 3, 13),
```

Comment:
You do a lot of repeated setup work throughout these tests. Would it be possible to consolidate this into a shared `before` block or something similar?